### ISSUE TITLE
Imported templates created by MacOS tar(1) break when editing in the webui

### ISSUE THREAD

**Issue by geiseri** [coder/coder#9565]
[TL;DR: ](https://invisible-island.net/autoconf/portability-tar.html#standard)"The nice thing about standards is that you have so many to choose from." -Andrew S. Tanenbaum

I imported a template as a tar file from my MacOS desktop. The tar file seemed to upload without issue and the template was built/deployed just fine.  Things went off the rails when I tried to edit it... 

```
create file "/home/coder/.cache/coder/provisioner-0/work/PaxHeader/currentdir" (mode -rwxr-xr-x): open /home/coder/.cache/coder/provisioner-0/work/PaxHeader/currentdir: no such file or directory
```

Some spelunking led me on a spiritual journey of tar 

![tar](https://imgs.xkcd.com/comics/tar.png).  

Nobody is innocent here so I will start with FreeBSD.  MacOS is it's more stylish cousin and as such inherited its `tar(1)` implementation.  Unlike GNU, FreeBSD likes to take the moral high road and thus decided to use the POSIX.1-2001/[PAX]( https://pubs.opengroup.org/onlinepubs/9699919799/utilities/pax.html#tag_20_92_13) for its [tar implementation](https://[www.libarchive.org](http://www.libarchive.org/)).  Unfortunately for them the Open Group would rather see the world burn than break any sort of interoperability with some dead UNIX from the 80s so pretty much every fallback scenario parses the record as just another file. Next we have the authors of `archive/tar`.  They hate other people's things and decided to rewrite their tar implementation from scratch and have not yet realized the hell they stepped into.   Then there is me.  I hate all nice things and decided to use as complicated of a setup as I can muster.  Finally, we have the coder team.  They desperately would like to have nice things. 

Unfortunately for them everyone in this story is conspiring against them.  My desktop is joined to Active Directory so my UID is above 2097151.  The libarchive devs are considerate about how they generate tar files in that they try to keep the output as universally portable as possible. (Do not confuse portable with compatible.) So, if the file that is being archived can be supported by basic ustar they will use that, otherwise they will bump up to PAX headers.  Once you enter the land of PAX though, things get complicated.  We are talking k-drama level complication.  If you look at the [~500 lines of if/else statements](https://github.com/libarchive/libarchive/blob/master/libarchive/archive_read_support_format_tar.c#L1586C1-L2202C1) you can get a taste for it if you don’t want to suffer through a season of 'Boys Over Flowers'.  I think a tar file with PAX headers confuses something in the chain (possibly in [mergePax](https://cs.opensource.google/go/go/+/refs/tags/go1.21.0:src/archive/tar/reader.go;l=260)) because PAX headers are completely valid files if you don't see them coming. When you view the template in the source you see this:

![image](https://github.com/coder/coder/assets/2265693/80880393-540a-403d-a517-cac7f6ad751a)

For reference GNU tar produces this:
![image](https://github.com/coder/coder/assets/2265693/0390b85a-7cfa-49f2-84d5-e5d009e3f957)

Strangely though when I looked at [archive.go](https://github.com/coder/coder/blob/main/provisionersdk/archive.go) it shouldn't care because it would have written the file on untar, or never looked for it on tar.  With this, I realized enlightenment was unattainable and decided to just rant my findings here.

The only ~constructive~ ideas I could come up with on how to address this. Wrap `libarchive` or one of the other complete implementations.  Have the authors of `archive/tar` handle that header since it looks like they explicitly hardcode some other popular PAX headers.  Ignore any directory name that starts with `PaxHeader`.  Document that the tar format should be in `ustar` when using MacOS.  Or just ban me from computers so the coder team can have one less reason we can’t have nice things.

**Comment by BrunoQuaresma** [comment]
Related to https://github.com/coder/coder/issues/11687

**Comment by geiseri** [comment]
Using `tar(1)` or `zip(1)` on MacOS?

**Comment by spikecurtis** [comment]
Thanks for the issue, and making it a fun read.

I think the least we can do is document that people should stick to `ustar` format. 

Wrapping some other library, like libarchive is probably not in the cards. We functionally just want a single blob that contains the template files, and don't really _want_ to have to care about uids, guids, etc., so handling pax format probably doesn't add any new capabilities in what you can express.

@geiseri if you could attach a tar file that reproduces the problem, we can step through _our_ go code to see if there is anything reasonable we could be doing differently to work around the issue.

Also let us know if using `zip` on your setup works correctly.  Ultimately, our goal is to make the Coder API usable in different environments without jumping through a bunch of hoops, so if you can substitute `zip` for `tar` and get an OK experience that's a win.

**Comment by spikecurtis** [comment]
We've merged docs changes in #14152 

I'm going to close this issue, but @geiseri please re-open if you're able to provide us with a tar file that reproduces your problem. I totally understand if you've moved on to other things and it's not so easy.
